import sys
from datetime import datetime
import pandas as pd

configfile: 'config/config.yaml'

##### load rules #####
include: 'rules/count.smk'
include: 'rules/vdj.smk'
include: 'rules/wf-single-cell.smk'

##### main rules #####

df=(pd.read_table(config['samples'], dtype={"sample_id": str, "sample_type": str, "sample_fastq_dir":str})
	.set_index("sample_id", drop=False)
	.sort_index()
)

df_illumina_gex =df[(df == 'illumina_gex').any(axis=1)]
df_illumina_vdj =df[(df == 'illumina_vdj').any(axis=1)]
df_ont = df[(df == 'ont').any(axis=1)]

rule all:
	input:
		expand('results/ill/cellranger_count/{sample}/outs/possorted_genome_bam.bam', sample=df_illumina_gex['sample_id'].tolist()),
		expand('results/ont/wf-single-cell/{sample}/{sample}/{sample}.gene_expression.counts.tsv', sample=df_ont['sample_id'].tolist()),
		expand('results/ont/wf-single-cell/{sample}/{sample}/10x-genes/matrix.mtx.gz', sample=df_ont['sample_id'].tolist()),
		expand('results/ont/wf-single-cell/{sample}/{sample}/10x-transcripts/matrix.mtx.gz', sample=df_ont['sample_id'].tolist()),
		expand('results/ill/cellranger_vdj/{sample}/outs/consensus.bam', sample=df_illumina_vdj['sample_id'].tolist())	

rule cellranger_count:
	input:
		expand('results/ill/cellranger_count/{sample}/outs/possorted_genome_bam.bam', sample=df_illumina_gex['sample_id'].tolist())

rule cellranger_vdj:
	input:
		expand('results/ill/cellranger_vdj/{sample}/outs/consensus.bam', sample=df_illumina_vdj['sample_id'].tolist())

rule wf_single_cell:
	input:
		expand('results/ont/wf-single-cell/{sample}/{sample}/{sample}.gene_expression.counts.tsv', sample=df_ont['sample_id'].tolist()),
		expand('results/ont/wf-single-cell/{sample}/{sample}/10x-genes/matrix.mtx.gz', sample=df_ont['sample_id'].tolist()),
		expand('results/ont/wf-single-cell/{sample}/{sample}/10x-transcripts/matrix.mtx.gz', sample=df_ont['sample_id'].tolist())
